app/src/main/java/org/astraea/metrics/jmx/BeanObject.java

```diff
@@ -0,0 +1,70 @@
+package org.astraea.metrics.jmx;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.management.Attribute;
+import javax.management.AttributeList;
+
+public class BeanObject {
+  private final String domainName;
+  private final Map<String, String> properties;
+  private final Map<String, Object> attributes;
+
+  public BeanObject(
+      String domainName, Map<String, String> properties, Map<String, Object> attributes) {
+    this.domainName = domainName;
+    this.properties = new HashMap<>(properties); // making defensive copy
+    this.attributes = new HashMap<>(attributes); // making defensive copy
+  }
+
+  public BeanObject(String domainName) {
+    this.domainName = domainName;
+    this.properties = new HashMap<>();
+    this.attributes = new HashMap<>();
+  }
+
+  public BeanObject selectProperty(String key, String value) {
+    this.properties.put(key, value);
+    return this;
+  }
+
+  public BeanObject fetchAttribute(String attributeName) {
+    this.attributes.put(attributeName, null);
```

Comment:
If we don't plan to update inner members in the future, making object be immutable is always the better pattern. Another way is to return a new object when users try to update the object.

Reply:
This is a good idea, but ``BeanObject::updateAttributeValue`` can be used to update its internal state. If the update happens frequently, this will create a lot of objects.
This is the reason I didn't make it immutable.

But still, your idea gives me some inspiration. Maybe we can make the object immutable but make an exception for ``BeanObject::updateAttributeValue``. Any invocation of this method won't create a new object. This can avoid the above problem.

After all, that updated attribute value is a snapshot from the remote server, not a serious internal state to this class, doing so should be ok.

Reply:
> BeanObject::updateAttributeValue can be used to update its internal state. If the update happens frequently, this will create a lot of objects.
This is the reason I didn't make it immutable.

That is a good reason if we will have such use case in the future. 

> But still, your idea gives me some inspiration. Maybe we can make the object immutable but make an exception for BeanObject::updateAttributeValue. Any invocation of this method won't create a new object. This can avoid the above problem.

That is weird to me. If the object is immutable, we should not add any method which can "update" inner state.

> that updated attribute value is a snapshot from the remote server, not a serious internal state to this class, doing so should be ok.

Normally, the "snapshot" is immutable so I prefer to have an immutable object to represent "snapshot".

At any rate, it is ok to keep the "update" method if you assume that we will have such use cases in the future.

Reply:
> That is weird to me

I feel weird about this too. 

I got a better idea, maybe we revise the design of ``BeanObject``.

Change it from 

```java
public class BeanObject {
  private final String domainName;
  private final Map<String, String> properties;
  private final Map<String, Object> attributes;
  // ...
```

to

```java
public class BeanObject {
  private final String domainName;
  private final Map<String, String> properties;
  private final Set<String> attributes;
  // ...
```

The above code separates the attribute value from ``BeanObject`` since the value is the part that tends to change.

Reply:
> The above code separates the attribute value from BeanObject since the value is the part that tends to change.

Pardon me, I still fail to catch the point about "why we need to update attributes"? If we get a good use case of updating MBean attributes, it is fine to add any update method for it.

Reply:
Also rename ``BeanObject`` to ``BeanStruct`` or something, since now the object doesn't carry those actual attribute values.

Reply:
>  now the object doesn't carry those actual attribute values.

why? MBean attributes are used to carry data also. For example: kafka topic has following value stored by attributes: `MeanRate` and `RateUnit`. It seems to me the attributes should be a part of "snapshot" and should carry the "actual" data of remote server.

Reply:
> Pardon me, I still fail to catch the point about "why we need to update attributes"? If we get a good use case of updating MBean attributes, it is fine to add any update method for it.

Originally I did this for performance reasons(avoid create many objects). I just realize if we query so often that this becomes a problem, the Kafka metric server should die before our client.

Maybe I am doing "premature optimization".

> >Also rename BeanObject to BeanStruct or something, since now the object doesn't carry those actual attribute values.

Oops, this is an accident, when I posting the comment I didn't see you respond already.

> MBean attributes are used to carry data also. For example: kafka topic has following value stored by attributes: MeanRate and RateUnit. It seems to me the attributes should be a part of "snapshot" and should carry the "actual" data of remote server.

You are right, I remember these MBean thing doesn't have a concrete schema, keep the data with its schema is a better way to handle it.

Reply:
I think to stick to the original design, but make the ``BeanObject`` completely immutable. should be ok.

Reply:
> I think to stick to the original design, but make the BeanObject completely immutable. should be ok.

Please add the comment to BeanObject to say why it should be immutable now.

Reply:
> Please add the comment 

Do you mean a Github comment at that line of code or Javadoc on ``BeanObject``?

Reply:
I add the Javadoc here https://github.com/skiptests/astraea/pull/15/commits/78f6bcce1038b1ea4d18e9c71673e38f45fb7bbd.
I am not sure the intention behind this, maybe it is used to remind other maintainers. So I write some notice on these classes